FinancialTracker/Firebase.swift

```diff
@@ -15,106 +15,165 @@ enum FirebaseError: Error {
                 return
             }
 
             guard let result = result else {
-                completionHandler(FirebaseError.unknown, false)
+                completionHandler(FirebaseError.unknown, nil)
                 return
             }
 
-            let firstNameKey = UserDataProperty.firstName.rawValue
-            let lastNameKey = UserDataProperty.lastName.rawValue
-            let uidKey = UserDataProperty.uid.rawValue
-            let data = [firstNameKey: firstName, lastNameKey: lastName, uidKey: result.user.uid]
+            let firstNameKey = User.CodingKeys.firstName.rawValue
+            let lastNameKey = User.CodingKeys.lastName.rawValue
+            let uidKey = User.CodingKeys.uid.rawValue
+            let emailKey = User.CodingKeys.email.rawValue
+            let data = [firstNameKey: firstName, lastNameKey: lastName, emailKey: email, uidKey: result.user.uid]
             
-            self.database.collection(DBCollectionKey.users.rawValue).document(result.user.uid).setData(data) { error in
+            self.firestore.collection(DBCollectionKey.users.rawValue).document(result.user.uid).setData(data) { error in
                 if error != nil {
-                    completionHandler(FirebaseError.database(error), false)
+                    completionHandler(FirebaseError.database(error), nil)
                     return
                 }
-                completionHandler(nil, true)
+                
+                self.user = User(firstName: firstName, lastName: lastName, email: email, uid: result.user.uid)
+                completionHandler(nil, self.user)
             }
         }
     }
 
-    func signIn(email: String, password: String, completionHandler: @escaping (FirebaseError?, Bool) -> Void) {
+    func signIn(email: String, password: String, completionHandler: @escaping (FirebaseError?, User?) -> Void) {
         auth.signIn(withEmail: email, password: password) { _, error in
-
             guard error == nil else {
-                completionHandler(FirebaseError.auth(error), false)
+                completionHandler(FirebaseError.auth(error), nil)
                 return
             }
-            completionHandler(nil, true)
-        }
-    }
-    
-    func signOut(completionHandler: @escaping (FirebaseError?, Bool) -> Void) {
-        do {
-            try auth.signOut()
-        } catch let signOutError {
-            completionHandler(FirebaseError.signOut(signOutError), false)
+
+            self.getCurrentUserData { firebaseError, user in
+                switch firebaseError {
+                case .database(let error):
+                    completionHandler(FirebaseError.database(error), user)
+                case .unknown:
+                    completionHandler(FirebaseError.unknown, user)
+                case .none:
+                    completionHandler(nil, user)
+                    self.user = user
+                default:
+                    break
```

Comment:
if you keep the `default` case, put an `assert` in it, else - go through all cases, we discussed this right?